extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -870,19 +890,47 @@ class LifecycleTask extends BackbeatTask {
      * @return {undefined}
      */
     _applyTransitionRule(params, log) {
+        // TODO: Bugfix: version id is not set in params.
         async.waterfall([
             next =>
                 this._getObjectMD(params, log, next),
-            (objectMD, next) =>
-                this._getTransitionActionEntry(params, objectMD, log, next),
-            (entry, next) =>
+            (objectMD, next) => {
+                // If transition is in progress, do not re-publishing entry
+                // to data-mover or cold-archive topic.
+                if (objectMD.getTransitionInProgress()) {
+                    return next(errorTransitionInProgress);
+                }
+                return this._getTransitionActionEntry(params, objectMD, log, (err, entry) =>
+                    next(err, entry, objectMD));
+            },
+            (entry, objectMD, next) =>
                 ReplicationAPI.sendDataMoverAction(
-                    this.producer, entry, log, next),
+                    this.producer, entry, log, err => next(err, entry, objectMD)),
+            (entry, objectMD, next) => {
+                // Update object metadata with "x-amz-scal-transition-in-progress"
+                // to avoid transitioning object a second time from a new batch.
+                // Only implemented for transitions to cold location.
+                const toLocation = entry.getAttribute('toLocation');
+                const locationConfig = locationsConfig[toLocation];
+                if (locationConfig && locationConfig.isCold) {
+                        objectMD.setTransitionInProgress(true);
+                        const putParams = {
+                            bucket: params.bucket,
+                            objectKey: params.objectKey,
+                            versionId: params.versionId,
+                            mdBlob: objectMD.getSerialized(),
+                        };
+
+                        return this._putObjectMD(putParams, log, next);
+                }
+
+                return process.nextTick(next);
```

Comment:
why `nextTick` here?

Reply:
For the waterfall to behave the way it should.